### DISCUSSION TITLE
Pi Pico W, will only serve a single request and then stops

### DISCUSSION THREAD

**Discussion by adrianblakey** [miguelgrinberg/microdot#166] [General]
I have implemented the "hello world" with Debug True, I successfully connect to my WiFi network, point a browser at the IP address of 192.168.178.126 and it logs: "Starting threaded server on 0.0.0.0:80...", "GET / 200" and I see "Hello, from Pico"

If I resubmit the request from the browser, there is no response. Nothing is logged to show the request has been received, the browser times out - the site can not be reached.

I am running it from Thonny, Micro Python 2023/04/26 (and I have tried with the latest nightly build).

The only solution is to disconnect the Pi from USB, reconnect it and rerun.

Any ideas?

**Comment by miguelgrinberg** [comment]
Probably similar or at least related to https://github.com/miguelgrinberg/microdot/issues/89.

The issue is that some browsers have a performance optimization in which they open a connection and keep it open to have it ready when they need to send a new request to the server. This is problematic when using the synchronous version of Microdot because only one connection can be handled. In an attempt to outsmart browsers I have added a timeout on all connections, so that any connections that are open and don't send a request shortly after are closed. This appeared to have addressed the problem, in fact I haven't heard of any more issues with this since I added the timeout.

One thing you can try to determine if it is really the browser messing up with connections causing this is to start the server and then send a request from `curl` or any other simpler HTTP client. If you can continue sending curl requests then my guess is that this is still the browser trying to make more connections that it really needs.

The easiest way to address this problem is to use the async version of Microdot, which can handle multiple connections concurrently.

**Reply by adrianblakey** [reply]
Using the async version fixes it. TY

**Comment by bulletmark** [comment]
@miguelgrinberg given this constraint, I think you should consider dropping sync support and make Microdot an async only framework? It would significantly simplify the code, tests, and documentation. It would also simplify enhancements and reduce the time you spend on support. Largely it is inexperienced coders who start with a sync approach anyhow so you may also reduce the beginner Python questions you often have to deal with around here.

Microdot is a jewel in the MicroPython world so just thinking about how your time here can be made more effective.

**Reply by miguelgrinberg** [reply]
I have considered it, but you have to remember that the sync server supports threads for concurrency. The vast majority of boards do not have threads, but some do, and maybe more will appear in the future. And threads work really well on the Unix and Windows port of MicroPython, and of course also on CPython.
